csrc/cutlass/block_scaling.cpp

```diff
@@ -0,0 +1,326 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <cutlass/block_scaling.h>
+#include <exceptions.h>
+#include <fusion.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_nodes.h>
+#include <ir/utils.h>
+#include <type.h>
+
+namespace nvfuser {
+
+namespace {
+
+// These are the quantized types that make use of block scaling for which we
+// have established Fusion IR recipes.
+bool isBlockScaledDtype(const DataType& dtype) {
+  return dtype == DataType::Float4_e2m1fn || dtype == DataType::Float8_e4m3fn ||
+      dtype == DataType::Float8_e5m2;
+}
+
+} // namespace
+
+namespace cutlass_codegen {
+
+std::string BlockScaledOutputPattern::toString() const {
+  std::stringstream ss;
+  ss << "BlockScaledOutputPattern: {\n";
+  ss << "  unquantized_output: " << unquantized_output->toString() << "\n";
+  ss << "  quantized_output: " << quantized_output->toString() << "\n";
+  ss << "  block_scale_factors: " << block_scale_factors->toString() << "\n";
+  ss << "  block_size: " << block_size << "\n";
+  ss << "}\n";
+  return ss.str();
+}
+
+// This matches a standard block scaling pattern expressed in Fusion IR.
+//
+// Block scaling quantizes a tensor by dividing it by per-block scale factors
+// that are computed from the absolute maximum values within each block.
+//
+// Pattern diagram (without global scale):
+//
+//          unquantized_output (high precision)
+//                   |
+//       Reshape (split by block_size)
+//                   |
+//           data_hp_reshaped
+//           /            \
+//          /              \
+//         /                \.
+//        /                Abs
+//       |                   |
+//       |             Max (reduction)
+//       |                   |
+//       |            Div (by constant)
+//       |                   |
+//       |             [Optional: Clamp]
+//       |                   |
+//       |             Cast (to FP8)
+//       |                   |
+//       |           block_scale_fp8 --> OUTPUT (FP8 scales)
+//       |                   |
+//       |             Cast (to FP32)
+//       |                   |
+//       |           Broadcast/Unsqueeze
+//       |                   |
+//       +--------+----------+
+//                |
+//               Div
+//                |
+//           data_scaled
+//                |
+//         [Optional: Clamp]
+//                |
+//        Cast (to low precision)
+//                |
+//         [Optional: Reshape]
+//                |
+//         quantized_output --> OUTPUT (quantized)
```

Comment:
```suggestion
#if defined(__clang__) || defined(__GNUC__)
#pragma GCC diagnostic pop
#endif
```
**[Suggested by AI]** Summary of the fix:

• The line ​#pragma GCC diagnostic pop​ is now wrapped in an ​#if defined(__clang__) || defined(__GNUC__)​ … ​#endif​ block.  
• This means the “pop” directive is only emitted when the compiler is Clang or GCC, i.e., compilers that understand ​#pragma GCC diagnostic pop​.  
• Other compilers (e.g., MSVC) will now ignore this section, eliminating the unknown-pragma build error that was previously generated.